node/src/components/contract_runtime.rs

```diff
@@ -425,27 +427,43 @@ where
                     );
                     trace!(?result, "execute block response");
                     responder.respond(result).await
-                }
+                })
                 .ignore()
             }
             ContractRuntimeRequest::EnqueueBlockForExecution {
                 finalized_block,
                 deploys,
             } => {
                 info!(?finalized_block, "enqueuing finalized block for execution");
-                if self.execution_pre_state.next_block_height == finalized_block.height() {
-                    self.execute_finalized_block(
-                        effect_builder,
-                        self.protocol_version,
-                        self.execution_pre_state.clone(),
-                        finalized_block,
-                        deploys,
+                let mut effects = Effects::new();
+                let engine_state = Arc::clone(&self.engine_state);
+                let metrics = Arc::clone(&self.metrics);
+                let exec_queue = Arc::clone(&self.exec_queue);
+                let execution_pre_state = Arc::clone(&self.execution_pre_state);
+                let protocol_version = self.protocol_version;
+                if self.execution_pre_state.lock().unwrap().next_block_height
```

Comment:
We should probably `fatal!` here...

Reply:
I think it did happen to us in the past that two _finalized events_ `A1, A2` arrived in the reversed order `A2, A1. This ensures that we recover from that situation correctly.

Reply:
It's considered "normal" to unwrap a locked `std::sync::Mutex`, because the only time it will panic is if another thread panicked while holding the Mutex. Therefore if we handle other errors within the lifetime of the `MutexGuard`, this cannot occur. I'm happy to revisit but I think this probably obviated by a follow-up we're already planning.